### PR TITLE
Fetch

### PR DESCRIPTION
**Pull request by dobe**

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/JobIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                .build();
+    }
+
+    public JobIntegrationTest() {
+        // ensure that the client node is used as handler and has no collectphase
+        super(new SQLTransportExecutor(
+                new SQLTransportExecutor.ClientProvider() {
+                    @Override
+                    public Client client() {
+                        return internalCluster().client("node_s1");
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/JobIntegrationTest.java]: would it be possible to use .clientNodeClient() instead?

File: sql/src/main/java/io/crate/metadata/Routing.java [hunk 2]
```diff
@@ -15,7 +15,80 @@
...
+        private Map<String, Map<String, List<Integer>>> result;
+
+        @Override
+        public boolean visitLocations(Map<String, Map<String, List<Integer>>> locations) {
+            result = new HashMap<>(locations.size());
+            return true;
+        }
+
+        @Override
+        public boolean visitShard(String nodeId, String index, Integer shardId) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Routing.java]: seems like this visitShard is never used

File: sql/src/main/java/io/crate/planner/Planner.java [hunk 3]
```diff
@@ -109,59 +165,91 @@ public UUID jobId() {
...
+                routing = tableInfo.getRouting(where, preference);
+            } else {
+                for (TableRouting existing : existingRoutings) {
+                    assert preference == null || preference.equals(existing.preference);
+                    if (Objects.equals(existing.where, where)) {
+                        return existing.routing;
+                    }
+                }
+                // ensure all routings of this table are allocated
+                for (TableRouting existingRouting : existingRoutings) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/Planner.java]: the stuff here, including the rebuildRouting and the whole NodeRewritingVisitor doesn't have any coverage at all. Can we either write tests for it or replace it with an UnsupportedOperationException until it is actually needed?

File: sql/src/main/java/io/crate/planner/projection/FetchProjection.java [hunk 4]
```diff
@@ -177,14 +173,14 @@ public void readFrom(StreamInput in) throws IOException {
         }
 
         int numJobSearchContextIdToNode = in.readVInt();
-        jobSearchContextIdToNode = new IntObjectOpenHashMap<>(numJobSearchContextIdToNode);
+        readerNodes = new IntObjectOpenHashMap<>(numJobSearchContextIdToNode);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/projection/FetchProjection.java]: do we have a streaming test for the FetchProjection?

File: sql/src/main/java/io/crate/planner/Planner.java (hunk 3 above)
Comment (seut) [inline sql/src/main/java/io/crate/planner/Planner.java]: rebuild stuff is removed, added planner unit test to cover remaining

File: sql/src/main/java/io/crate/planner/projection/FetchProjection.java (hunk 4 above)
Comment (seut) [inline sql/src/main/java/io/crate/planner/projection/FetchProjection.java]: added a streaming test for the FetchProjection

**Comment by mfussenegger** [comment]
ack

**Pull request merged**